### PR TITLE
Make flatbuffers optional

### PR DESCRIPTION
**Pull request by olgavrou**
Making flatbuffers optional. ON by default.

Will update documentation but when building via cmake can be turned off by setting `-DBUILD_FLATBUFFERS=OFF` and on windows via visual studio will need to remove the pre-processor definition `BUILD_FLATBUFFERS`

### CODE REVIEW

File: test/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -35,13 +33,14 @@ add_executable(vw-unit-test.out
   test_common.cc
   test_common.h
   tokenize_tests.cc
```
Comment (olgavrou) [inline test/unit_test/CMakeLists.txt:35]: the removed files here are duplicates

File: test/unit_test/unit_test.vcxproj.filters [hunk 2]
```diff
@@ -81,9 +78,6 @@
     <ClCompile Include="options_boost_po_test.cc">
       <Filter>Source Files</Filter>
     </ClCompile>
-    <ClCompile Include="options_test.cc">
```
Comment (olgavrou) [inline test/unit_test/unit_test.vcxproj.filters:84]: duplicate

File: CMakeLists.txt [hunk 3]
```diff
@@ -197,7 +198,10 @@ endif()
 # Align and foreach are also required, for some reason they can't be specified as components though.
 find_package(Boost REQUIRED COMPONENTS program_options system)
 find_package(ZLIB REQUIRED)
-find_package(flatbuffers REQUIRED)
+if(BUILD_FLATBUFFERS)
+  find_package(flatbuffers REQUIRED)
+  add_definitions(-DBUILD_FLATBUFFERS)
```
Comment (jackgerrits) [inline CMakeLists.txt:203]: Use `target_compile_definitions` instead

File: test/unit_test/unit_test.vcxproj [hunk 4]
```diff
@@ -35,7 +35,7 @@
     <ClCompile Include="error_test.cc" />
     <ClCompile Include="example_header_test.cc" />
     <ClCompile Include="explore_test.cc" />
-    <ClCompile Include="flatbuffer_parser_test.cc" />
+    <ClCompile Include="flatbuffer_parser_test.cc" Condition="Exists('$(VcpkgIntegration)')" />
```
Comment (jackgerrits) [inline test/unit_test/unit_test.vcxproj:38]: Is it possible to condition this on `BUILD_FLATBUFFERS` or some env var?

File: CMakeLists.txt (hunk 3 above)
Comment (olgavrou) [reply CMakeLists.txt]: Done

File: test/unit_test/unit_test.vcxproj (hunk 4 above)
Comment (olgavrou) [reply test/unit_test/unit_test.vcxproj]: I tried it out

File: CMakeLists.txt [hunk 5]
```diff
@@ -250,7 +253,7 @@ add_subdirectory(cluster)
 add_subdirectory(library)
 add_subdirectory(vowpalwabbit)
 
-if (BUILD_FLATBUFFER_UTILS)
+if (BUILD_FLATBUFFER_UTILS AND BUILD_FLATBUFFERS)
```
Comment (jackgerrits) [inline CMakeLists.txt:256]: Should we collapse both of these options into just `BUILD_FLATBUFFERS`? The main reason we have the option is to control whether we require the flatbuffer dependency to be present or not. If it's present I feel like having the target present for fb utils is not an issue

**Review by jackgerrits — APPROVED** [review]

File: CMakeLists.txt (hunk 5 above)
Comment (olgavrou) [reply CMakeLists.txt:256]: Yes good point, will do

**Pull request merged**